app/scripts/metamask-controller.js

```diff
@@ -4658,39 +4659,49 @@ export default class MetamaskController extends EventEmitter {
       // Scan accounts until we find an empty one
       const chainId = this.#getGlobalChainId();
       const ethQuery = new EthQuery(this.provider);
-      const accounts = await this.keyringController.getAccounts();
-      let address = accounts[accounts.length - 1];
 
-      for (let count = accounts.length; ; count++) {
-        const balance = await this.getBalance(address, ethQuery);
+      await this.keyringController.withKeyring(
+        { type: KeyringTypes.hd },
+        async (hdKeyring) => {
+          let balance = '0x0';
+          do {
+            const [address] = hdKeyring.addAccounts(1);
 
-        if (balance === '0x0') {
-          // This account has no balance, so check for tokens
-          await this.tokenDetectionController.detectTokens({
-            chainIds: [chainId],
-            selectedAddress: address,
-          });
+            try {
+              balance = await this.getBalance(address, ethQuery);
+            } catch (error) {
+              // Errors are gracefully handled so that `withKeyring`
+              // will not rollback the primary keyring, and accounts
+              // created in previous loop iterations will remain in place.
```

Comment:
```suggestion
              // Errors are gracefully handled so that `withKeyring`
              // will not rollback the primary keyring, and accounts
              // created in previous loop iterations will remain in place.
              // The balance is set to zero in case of previous iterations,
              // so that token balances are checked anyway.
              balance = '0x0';
```